app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ImportExportApplicationService.java

```diff
@@ -0,0 +1,576 @@
+                        applicationJson.setDatasourceList(datasourceList);
+                        return newActionRepository.findByApplicationId(applicationId);
+                    })
+                    .flatMap(Flux::collectList)
+                    .map(newActionList -> {
+                        Set<String> concernedDBNames = new HashSet<>();
+                        newActionList.forEach(newAction -> {
+                            newAction.setPluginId(pluginMap.get(newAction.getPluginId()));
+                            newAction.setOrganizationId(null);
+                            newAction.setPolicies(null);
+                            newAction.setApplicationId(applicationName);
+                            //Collect Datasource names to filter only required datasources
+                            if (newAction.getPluginType() == PluginType.DB || newAction.getPluginType() == PluginType.API) {
+                                concernedDBNames.add(mapDatasourceIdToNewAction(newAction.getPublishedAction(), datasourceMap));
+                                concernedDBNames.add(mapDatasourceIdToNewAction(newAction.getUnpublishedAction(), datasourceMap));
+                            }
+                            if (newAction.getUnpublishedAction() != null) {
+                                ActionDTO actionDTO = newAction.getUnpublishedAction();
+                                actionDTO.setPageId(newPageIdMap.get(actionDTO.getPageId()));
+                            }
+                        });
+                        applicationJson
+                                .getDatasourceList()
+                                .removeIf(datasource -> !concernedDBNames.contains(datasource.getName()));
+
+                        return newActionList;
+                    })
+                    .map(newActionList -> {
+                        applicationJson.setActionList(newActionList);
+                        //Only export those datasources which are used in the app instead of org level
+                        Map<String, DecryptedSensitiveFields> decryptedFields = new HashMap<>();
+                        applicationJson.getDatasourceList().forEach(datasource -> {
+                            decryptedFields.put(datasource.getName(), getDecryptedFields(datasource));
+                            datasource.setId(null);
+                            datasource.setOrganizationId(null);
+                            datasource.setPluginId(pluginMap.get(datasource.getPluginId()));
+                            if (datasource.getDatasourceConfiguration() != null) {
+                                datasource.getDatasourceConfiguration().setAuthentication(null);
+                            }
+                        });
+                        applicationJson.setDecryptedFields(decryptedFields);
+                        return Mono.just(applicationJson);
```

Comment:
Using `Mono.just` in the return value of the callback to `.map` will cause double wrapping. You should be fine without it here.

```suggestion
                        return applicationJson;
```

Reply:
Updated code pushed!